# Part 1 - Data Loading

In [25]:
import pandas as pd
import numpy as np
import requests # fetch HTML contents from website
from bs4 import BeautifulSoup

## Scrape HTML Table Cells:

In [43]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
#Create a handle, page, to handle the contents of the website
res = requests.get(url)
#Parse data that are stored between <tr>..</tr> of HTML
soup = BeautifulSoup(res.content,'lxml')
table = soup.find_all('table')[0] 
table_rows = table.find_all('tr')

## Store Toronto Postal Code (FSA) Details if Borough is assigned:

In [54]:
res = []
for row in table_rows:
    cols = row.find_all('td') # get columns for each row
    if len(cols) == 3 and cols[1].text.strip() != "Not assigned": # verify correct no. columns and that Borough has a valid value assigned
            record = [row.text.strip() for row in cols if row.text.strip()]
            #print(row)
            if record:
                res.append(record) # add valid postal code record to the result list

## Create Pandas DataFrame and Update Table Header Columns:

In [55]:
# convert result list as dataframe and update column names
df = pd.DataFrame(res, columns=["Postcode","Borough","Neighbourhood"])

## Validate DataFrame Structure and View Records:

In [57]:
print(df.head(10))
print("Toronto FSA DataFrame Shape: ", df.shape)

  Postcode           Borough     Neighbourhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M5A  Downtown Toronto       Regent Park
4      M6A        North York  Lawrence Heights
5      M6A        North York    Lawrence Manor
6      M7A      Queen's Park      Not assigned
7      M9A         Etobicoke  Islington Avenue
8      M1B       Scarborough             Rouge
9      M1B       Scarborough           Malvern
Toronto FSA DataFrame Shape:  (211, 3)


Result shows the dataframe is created with **211** Postal Code records from the Wikipedia link *(which matches records found on Wikipedia)*.

## Set Neighbourhood value as the Borough if not assigned:

In [58]:
df.loc[df.Neighbourhood == 'Not assigned', 'Neighbourhood'] = df.Borough  

In [59]:
df.head(10)

Postcode           Borough     Neighbourhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M5A  Downtown Toronto       Regent Park
4      M6A        North York  Lawrence Heights
5      M6A        North York    Lawrence Manor
6      M7A      Queen's Park      Queen's Park
7      M9A         Etobicoke  Islington Avenue
8      M1B       Scarborough             Rouge
9      M1B       Scarborough           Malvern

Result shows the Neighbourhood value for Postal Code M7A is updated from ***Not assigned*** to the corresponding Borough, ***Queen's Park*** as expected.

## Remove duplicate rows for Postal code by combining records with different Neighbourhoods for the same Postcode and Borough:

In [87]:
grouped_df = df.groupby(['Postcode','Borough'], as_index=False, sort=False).agg(','.join)
grouped_df.head()
print("Shape of Formatted Postcode DataFrame: ", grouped_df.shape) # Records reduced from 221 to 103 with Neighbourhood grouped listing

Shape of Formatted Postcode DataFrame:  (103, 3)


# Part 2 - Add Latitude and Longitudes for Postal Codes:

In [83]:
latlong_df = pd.read_csv("http://cocl.us/Geospatial_data/Geospatial_Coordinates.csv").sort_values(by=['Postal Code'])

In [88]:
latlong_df.columns = ['Postcode','Latitude','Longitude'] # renaming "Postal Code" as "Postcode" to match dataframe from Part 1
postal_geo_df = pd.merge(grouped_df, latlong_df, on='Postcode')
postal_geo_df.head(5)

Postcode           Borough                    Neighbourhood   Latitude  \
0      M3A        North York                        Parkwoods  43.753259   
1      M4A        North York                 Victoria Village  43.725882   
2      M5A  Downtown Toronto         Harbourfront,Regent Park  43.654260   
3      M6A        North York  Lawrence Heights,Lawrence Manor  43.718518   
4      M7A      Queen's Park                     Queen's Park  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

# Part 3 - Clustering

In [104]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values


#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

### Get Latitude and Longitude for Toronto:

In [102]:
toronto_df = postal_geo_df[postal_geo_df['Borough'].str.contains('Toronto')] #Analyze Toronto Boroughs only
toronto_df.columns = ['Postcode','Borough','Neighborhood','Latitude','Longitude']

In [105]:

address = 'Toronto, ON'

geolocator = Nominatim(user_agent="tn_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


### Create map of Toronto with neighborhoods displayed:

In [107]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_df['Latitude'], toronto_df['Longitude'], toronto_df['Borough'], toronto_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Explore and Segment Toronto Neighborhoods using Foursquare API  
#### Foursquare Credentials

In [108]:
CLIENT_ID = 'M5RVHUFECMFIPNYREQQN4K5SFKTO0QNLVG5O5IEDVHJQYFUE' # your Foursquare ID
CLIENT_SECRET = 'HRIPJWCCZPNBZJ0OY3KIEZB4QXIX4410XBKW5B5FWGQH5DER' # your Foursquare Secret
VERSION = '20190515' # Foursquare API version

#### Explore Neighborhoods in Toronto

In [111]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [112]:
toronto_venues = getNearbyVenues(names=toronto_df['Neighborhood'],
                                   latitudes=toronto_df['Latitude'],
                                   longitudes=toronto_df['Longitude']
                                  )

Harbourfront,Regent Park
Ryerson,Garden District
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Adelaide,King,Richmond
Dovercourt Village,Dufferin
Harbourfront East,Toronto Islands,Union Station
Little Portugal,Trinity
The Danforth West,Riverdale
Design Exchange,Toronto Dominion Centre
Brockton,Exhibition Place,Parkdale Village
The Beaches West,India Bazaar
Commerce Court,Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North,Forest Hill West
High Park,The Junction South
North Toronto West
The Annex,North Midtown,Yorkville
Parkdale,Roncesvalles
Davisville
Harbord,University of Toronto
Runnymede,Swansea
Moore Park,Summerhill East
Chinatown,Grange Park,Kensington Market
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Rosedale
Stn A PO Boxes 25 The Esplanade
Cabbagetown,St. James Town
First Canadian Place,Underground city


#### Unique Venue Categories Found:

In [113]:
print(toronto_venues.shape)
toronto_venues.head()
toronto_venues.groupby('Neighborhood').count() # Total Venues by Neighborhood
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

(1700, 7)
There are 236 uniques categories.


#### Analyze Neighborhoods

In [114]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()
toronto_onehot.shape

# group rows by neighborhood and get mean occurrence frequency for each category
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.shape

# view the top 5 venues by neighborhood
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
                 venue  freq
0          Coffee Shop  0.06
1                 Café  0.05
2      Thai Restaurant  0.04
3           Steakhouse  0.04
4  American Restaurant  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2          Restaurant  0.04
3  Seafood Restaurant  0.04
4              Bakery  0.04


----Brockton,Exhibition Place,Parkdale Village----
            venue  freq
0  Breakfast Spot  0.11
1            Café  0.11
2     Coffee Shop  0.11
3   Burrito Place  0.05
4             Bar  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.12
1         Yoga Studio  0.06
2       Garden Center  0.06
3          Comic Shop  0.06
4                Park  0.06


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0   Airport Service  0.20
1    Airport Loun

In [115]:
# sort venues
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [116]:
# top 10 venues for each neighborhood
num_top_venues = 10
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2         Brockton,Exhibition Place,Parkdale Village                  Café   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...       Airport Service   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café       Thai Restaurant            Steakhouse   
1          Cocktail Bar                  Café           Cheese Shop   
2        Breakfast Spot           Coffee Shop         Grocery Store   
3                Garden      Recording Studio         Auto Workshop   
4        Airport Lounge      Airport Terminal       Harbor / Marina   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   American Restaurant                   Bar          Burger Joint   
1        Farmers Market              Beer Bar            Steakhouse   
2    Italian Restaurant  Caribbean Restaurant               Stadium   
3            Skate Park         Burrito Place  Fast Food Restaurant   
4      Sculpture Garden              Boutique                 Plane   

  8th Most Common Venue   9th Most Common Venue 10th Most Common Venue  
0                   Gym                   Hotel                 Bakery  
1    Seafood Restaurant              Restaurant                 Bakery  
2                   Bar  Furniture / Home Store          Burrito Place  
3        Farmers Market                 Brewery             Restaurant  
4         Boat or Ferry            Airport Gate                Airport

### Cluster Neighborhoods

In [117]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

#### Neighborhood Clusters (k=5) and top 10 venues by neighborhood:

In [118]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postcode           Borough              Neighborhood   Latitude  Longitude  \
2       M5A  Downtown Toronto  Harbourfront,Regent Park  43.654260 -79.360636   
9       M5B  Downtown Toronto   Ryerson,Garden District  43.657162 -79.378937   
15      M5C  Downtown Toronto            St. James Town  43.651494 -79.375418   
19      M4E      East Toronto               The Beaches  43.676357 -79.293031   
20      M5E  Downtown Toronto               Berczy Park  43.644771 -79.373306   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
2                1           Coffee Shop                  Park   
9                1           Coffee Shop        Clothing Store   
15               1           Coffee Shop                 Hotel   
19               1     Health Food Store                   Pub   
20               1           Coffee Shop          Cocktail Bar   

   3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
2                   Café                   Pub                     Bakery   
9         Cosmetics Shop                  Café  Middle Eastern Restaurant   
15            Restaurant                  Café             Cosmetics Shop   
19           Music Venue           Wings Joint           Department Store   
20                  Café           Cheese Shop             Farmers Market   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
2     Mexican Restaurant        Breakfast Spot                      Theater   
9               Tea Room           Pizza Place          Sporting Goods Shop   
15        Breakfast Spot             Gastropub                       Bakery   
19  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   
20              Beer Bar            Steakhouse           Seafood Restaurant   

   9th Most Common Venue 10th Most Common Venue  
2         Chocolate Shop             Beer Store  
9    Japanese Restaurant   Fast Food Restaurant  
15    Italian Restaurant                  Diner  
19   Dumpling Restaurant             Donut Shop  
20            Restaurant                 Bakery

#### Visualize Clusters

In [119]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine Clusters  
Examine clusters to identify venue categories that distinguish each cluster. Assign a name to each cluster.

In [131]:
#Cluster 1
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, 'Cluster Name'] = "Outdoors/Nature" 

#Cluster 2
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, 'Cluster Name'] = "Cafe" 

#Cluster 3
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, 'Cluster Name'] = "Garden" 

#Cluster 4
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, 'Cluster Name'] = "Public Transportation"

#Cluster 5
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, 'Cluster Name'] = "Public Park" 
toronto_merged.head()

Postcode           Borough              Neighborhood   Latitude  Longitude  \
2       M5A  Downtown Toronto  Harbourfront,Regent Park  43.654260 -79.360636   
9       M5B  Downtown Toronto   Ryerson,Garden District  43.657162 -79.378937   
15      M5C  Downtown Toronto            St. James Town  43.651494 -79.375418   
19      M4E      East Toronto               The Beaches  43.676357 -79.293031   
20      M5E  Downtown Toronto               Berczy Park  43.644771 -79.373306   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
2                1           Coffee Shop                  Park   
9                1           Coffee Shop        Clothing Store   
15               1           Coffee Shop                 Hotel   
19               1     Health Food Store                   Pub   
20               1           Coffee Shop          Cocktail Bar   

   3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
2                   Café                   Pub                     Bakery   
9         Cosmetics Shop                  Café  Middle Eastern Restaurant   
15            Restaurant                  Café             Cosmetics Shop   
19           Music Venue           Wings Joint           Department Store   
20                  Café           Cheese Shop             Farmers Market   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
2     Mexican Restaurant        Breakfast Spot                      Theater   
9               Tea Room           Pizza Place          Sporting Goods Shop   
15        Breakfast Spot             Gastropub                       Bakery   
19  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   
20              Beer Bar            Steakhouse           Seafood Restaurant   

   9th Most Common Venue 10th Most Common Venue Cluster Name  
2         Chocolate Shop             Beer Store         Cafe  
9    Japanese Restaurant   Fast Food Restaurant         Cafe  
15    Italian Restaurant                  Diner         Cafe  
19   Dumpling Restaurant             Donut Shop         Cafe  
20            Restaurant                 Bakery         Cafe